.github/workflows/build-test.yaml

```diff
@@ -1,29 +1,246 @@
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: debos-builder
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
+          target: builder
 
       - name: Run unit tests
         run: |
           docker-compose -f docker/unit-tests.test.yml \
-            up --build --exit-code-from=sut
+            up --exit-code-from=sut
+
+  recipe-tests:
+    name: Run recipe tests
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: debos
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
 
       - name: Run test recipes on host
         run: |
           docker-compose -f docker/recipes.test.yml \
-            up --build --exit-code-from=sut
+            up --exit-code-from=sut
+
+  uml-tests:
+    name: Run UML tests
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: debos
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
 
       - name: Run test recipes using UML backend
         run: |
           docker-compose -f docker/recipes-test-uml.yml \
-            up --build --exit-code-from=sut
+            up --exit-code-from=sut
+
+  publish-github:
+    name: Publish to GHCR
+    needs:
+      - unit-tests
+      - recipe-tests
+      - uml-tests
+    if: github.event_name != 'pull_request'
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Extract Docker metadata
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: ${{ env.GITHUB_TAG }}
+          tags: |
+            "type=raw,value=latest,enable=${{ github.ref == format('refs/heads/{0}', github.event.repository.default_branch) }}"
```

Comment:
you probably want to set:
```
flavor: |
  latest = auto
```
to avoid generating latest on tagging?

Reply:
from what i have seen `docker/metadata-action` is quite fickle and only creates the latest tag for releases which are tagged in git which doesn't work for our workflow; we want any push to the default branch to also be tagged as latest.

I'll do some more testing on a fork though :-)

Reply:
sorry i meant set `latest = false`

But we should really move to pushing :latest on tags/releases indeed rather then on every push

Reply:
right that's a great way forward then - I will remove this hack then ;-)